### ISSUE TITLE
Speed up uploads by uploading multiple observation files per connection

### ISSUE THREAD

**Issue by Djfe** [mozilla/MozStumbler#1632]
Why does MozStumbler upload reports only in 200-300 byte chunks?
That causes more traffic than needed because of the additional overhead and it takes very long to establish so many connections (for example if you have 1000 reports or more in a queue)

it would be far more efficient to concat reports up to 500kb or even 1mb and then send these bigger chunks -> way faster and less overhead (because of fewer packets)

**Comment by hannosch** [comment]
Not sure this is related, but the service side had a problem for the past week. If any uploads contained cell data without any radioType field set, the service did accept and process the data, but failed in a final step. This caused a HTTP 500 error to be send back to Mozilla Stumbler. I believe it retries those uploads after some time, as 500 errors are usually transient server errors.

But that should have manifested in the upload queue never going down to zero. So this might be totally unrelated. But maybe this problem only manifested with this service side bug in place. I've hotfixed the production service earlier today, so things should return to normal again.

**Comment by Djfe** [comment]
yeah I noticed that, 71 entries remained in the queue, I was able to upload them today
and I noticed some 500s in the log
but the rest uploaded fine, it just took a long time and the log said that they were uploaded in just steps of a few hundred bytes which isn't much in my oppinion

**Comment by garvankeeley** [comment]
@Djfe How long is the upload taking? 

I would be curious to know if doubling or tripling the size of the POST actually makes an order of magnitude speed difference. 
The current advantages, the POST is 1:1 with the file on disk, which makes the code simple, if a POST succeeds, we delete the file; also the larger the POST the more chance of failure.

**Comment by crankycoder** [comment]
I'm reluctant to see this change.  The total number of requests we're making is not that many - and the typical use case is that we're uploading on wifi networks where connections are generally pretty fast.

Once we start batching multiple reports from disk into a single network request, we need to start worrying about how to handle failed uploads for the network transaction.

Do we retransmit the entire batch?  I'm also concerned like @garvankeeley of network failures in the case of larger file uploads.

As this problem has only been reported recently and seems to be resolved as per https://github.com/mozilla/MozStumbler/issues/1632#issuecomment-96432186  - I'd rather we just leave things alone.

**Comment by gpoo** [comment]
I am unsure if related, but last week it took my Nexus 1 more than 8 hours to upload ~2000 records.  It was using Wifi, nothing but MozStumbler installed.  My Nexus 5 does it relatively quick, either by LTE or Wifi (although I have never had accumulated that much records to report)

**Comment by hannosch** [comment]
One more data point: Once the client has established the SSL connection to the service, it takes no more than 100ms to process each upload on the service end. 90th percentile is at 20ms. So I don't think there's anything wrong on the service end here. These times have been stable for a long while now.

**Comment by garvankeeley** [comment]
SSL connections do take much longer to establish than non-SSL, that could be contributing to the problem. 
If stumbler was to upload more per connection I think it could be coded such that there is no cost/risk to doing to, so I support the idea.

**Comment by Djfe** [comment]
well it happened, while riding a train -> the ssl connection needed to be reestablished all the time

@crankycoder why shouldn't we resend the whole batch
the only cases where http errors are send are, that there was a problem with the connection or on the server side
if the reports are faulty the server just throws them away anyways

simply concatenating reports and taking a look at the available memory should solve the problem
I think that the map itself takes up more ram than just a few reports (which are in the ram anyways when recorded in the same session so that they can be displayed on the map)

**Comment by hannosch** [comment]
Related to this, the upload code explicitly closes connections after a single batch upload for Android versions 4.0-4.3 / API versions 14-18 (https://github.com/mozilla/MozStumbler/blob/dev/libraries/stumbler/src/main/java/org/mozilla/mozstumbler/service/core/http/HttpUtil.java#L170).

In those versions the underlying libraries have a bug and you can't actually use keep-alive connections. Maybe there are workaround for this and on newer Android versions it should be possible to do keep-alive.
